### PR TITLE
[c10d] Move DDP broadcast coalesced to C++

### PR DESCRIPTION
**Pull request by goldsborough**
This PR depends on the tests added in #9670. It moves the first, tiny function from the c10d DDP to C++: `dist_broadcast_coalesced`. Let me know if ` torch/csrc/distributed/c10d/ddp.h` will be a good place to put these rewritten functions.

@pietern @teng-li @apaszke

### CODE REVIEW

File: torch/csrc/distributed/c10d/ddp.h
Comment (teng-li) [inline torch/csrc/distributed/c10d/ddp.h]: let's keep the same coding style in c10d, distBroadcastCoalesced, and the following like broadcastOptions.

**Review by teng-li — APPROVED** [review]
Looks good, please follow the c10d coding convention.

**Comment by goldsborough** [comment]
@apaszke @ezyang can I get codeowner approval, thanks

File: torch/csrc/distributed/c10d/ddp.h
Comment (ezyang) [inline torch/csrc/distributed/c10d/ddp.h]: Is there any particular reason for this to be inline?

**Review by ezyang — APPROVED** [review]

File: torch/csrc/distributed/c10d/ddp.h
Comment (apaszke) [inline torch/csrc/distributed/c10d/ddp.h]: It might actually be a better idea to first queue all the waits, and only then trigger the copies (you can accumulate the futures in a vector).

**Review by apaszke — APPROVED** [review]

File: torch/csrc/distributed/c10d/init.cpp
Comment (apaszke) [inline torch/csrc/distributed/c10d/init.cpp]: Why did those change to external includes? They are including code internal to this project

File: torch/csrc/distributed/c10d/ddp.h
Comment (goldsborough) [reply torch/csrc/distributed/c10d/ddp.h]: The next PR splits this file into a `.h` and `.cpp` so it solves this. For now, the function definition was in the header so I'll need `inline` or `static`

File: torch/csrc/distributed/c10d/init.cpp
Comment (goldsborough) [reply torch/csrc/distributed/c10d/init.cpp]: Do we have a guide for `"header"` being internal and `<header>` being external? It always seemed arbitrary to me :D From a C++ point of view, `"path/to/header"` means relative to the current file and `<path/to/header>` means relative to one of the include directories. Since we use absolute paths for all headers the angle brackets are more correct, but it's more of a stylistic issue than anything else

File: torch/csrc/distributed/c10d/ddp.h
Comment (goldsborough) [reply torch/csrc/distributed/c10d/ddp.h]: Good idea, I'll do that

Comment (goldsborough) [reply torch/csrc/distributed/c10d/ddp.h]: This should be a decent improvement since chunks are now broadcast in parallel, while the code used to be all serial?

**Comment by goldsborough** [comment]
@apaszke can you take a look at the new version?

**Review by teng-li — COMMENTED** [review]
The rest looks good to me.

Please run  clang-format-7 -i on both files before landing

File: torch/csrc/distributed/c10d/ddp.h
Comment (goldsborough) [reply torch/csrc/distributed/c10d/ddp.h]: sounds good

**Comment by goldsborough** [comment]
clang-formatted and re-ran tests on FAIR cluster. Landing

**Pull request closed without merging**
